File: ADE-stable-1.0/frontend/src/services/agent.service.ts
import { APIError } from '../types/error';
import {
  GenerationType,
  ReviewType,
  TestType,
  DocumentationType,
  CodeGenerationRequest,
  CodeGenerationResponse,
  CodeReviewRequest,
  CodeReviewResponse,
  TestGenerationRequest,
  TestGenerationResponse,
  DocumentationRequest,
  DocumentationResponse,
  AgentCapabilities,
} from '../types/agent.types';
import { MonitoringService } from './monitoring.service';

export interface GenerationContext {
  project_id: string;
  language: string;
  framework?: string;
  requirements: string[];
  constraints?: string[];
  style_guide?: Record<string, any>;
  existing_code?: string;
  metadata?: Record<string, any>;
}

export interface GeneratedCode {
  id: string;
  type: GenerationType;
  content: string;
  language: string;
  file_path: string;
  dependencies?: string[];
  tests?: string[];
  documentation?: string;
  generated_at: string;
  metadata?: Record<string, any>;
}

export interface ReviewComment {
  id: string;
  line_number: number;
  type: string;
  message: string;
  suggestion?: string;
  severity: string;
  category: string;
  metadata?: Record<string, any>;
}

export interface CodeReview {
  id: string;
  project_id: string;
  file_path: string;
  review_types: ReviewType[];
  comments: ReviewComment[];
  summary: string;
  recommendations: string[];
  generated_at: string;
  metadata?: Record<string, any>;
}

export interface TestRequest {
  project_id: string;
  file_path: string;
  test_types: TestType[];
  coverage_target?: number;
  framework?: string;
  metadata?: Record<string, any>;
}

export interface TestCase {
  id: string;
  name: string;
  description: string;
  type: TestType;
  setup?: string;
  teardown?: string;
  test_code: string;
  expected_result: string;
  metadata?: Record<string, any>;
}

export interface TestSuite {
  id: string;
  project_id: string;
  file_path: string;
  test_types: TestType[];
  test_cases: TestCase[];
  coverage?: number;
  generated_at: string;
  metadata?: Record<string, any>;
}

export interface Documentation {
  id: string;
  project_id: string;
  doc_types: DocumentationType[];
  content: string;
  format: string;
  examples?: string[];
  generated_at: string;
  metadata?: Record<string, any>;
}

interface AgentContext {
  taskDescription?: string;
  progressStatus?: string;
  blockers?: string[];
  previousSolutions?: any[];
}

interface AgentSolution {
  agent: string;
  response: string;
  model: string;
  metrics: {
    latency: number;
    tokens: number;
  };
}

interface WorkflowGuidance {
  initiative_guidance: string;
  progress_assessment: string;
}

interface AgentResponse {
  workflow_guidance: WorkflowGuidance;
  agent_solutions: AgentSolution[];
  coordinator_result: string;
}

interface AgentConfig {
  name: string;
  role: string;
  model: string;
}

interface AgentList {
  solver_agents: AgentConfig[];
  workflow_agents: AgentConfig[];
  coordinator: AgentConfig;
}

class AgentServiceImpl {
  private readonly API_BASE_URL = process.env.REACT_APP_API_BASE_URL || 'http://localhost:8000';
  private readonly monitoring: MonitoringService;

  constructor(monitoring: MonitoringService) {
    this.monitoring = monitoring;
  }

  async processWithAgents(
    prompt: string,
    context: AgentContext = {}
  ): Promise<AgentResponse> {
    const startTime = performance.now();

    try {
      const response = await fetch(`${this.API_BASE_URL}/api/agents/process`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          prompt,
          context: {
            task_description: context.taskDescription || '',
            progress_status: context.progressStatus || 'Not started',
            blockers: context.blockers || [],
            previous_solutions: context.previousSolutions || [],
          },
        }),
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const data = await response.json();
      const latency = performance.now() - startTime;

      // Track performance metrics
      this.monitoring.trackPerformance({
        name: 'agent_processing',
        value: latency,
        tags: {
          num_agents: data.agent_solutions.length.toString(),
          prompt_length: prompt.length.toString(),
        }
      });

      return data;
    } catch (error) {
      // Track error
      this.monitoring.trackError({
        message: (error as Error).message,
        stack: (error as Error).stack,
        severity: 'error',
        tags: {
          context: 'agent_processing'
        }
      });
      throw error;
    }
  }

  async listAgents(): Promise<AgentList> {
    try {
      const response = await fetch(`${this.API_BASE_URL}/api/agents/agents`);

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const data = await response.json();
      return data;
    } catch (error) {
      this.monitoring.trackError({
        message: (error as Error).message,
        stack: (error as Error).stack,
        severity: 'error',
        tags: {
          context: 'list_agents'
        }
      });
      throw error;
    }
  }

  async trackAgentMetrics(metrics: {
    agent: string;
    latency: number;
    tokens: number;
  }): Promise<void> {
    this.monitoring.trackPerformance({
      name: 'agent_metrics',
      value: metrics.latency,
      tags: {
        agent: metrics.agent,
        tokens: metrics.tokens.toString(),
      }
    });
  }
}

export const agentService = new AgentServiceImpl(MonitoringService.getInstance());

export class AgentService {
  private baseUrl: string;
  private token: string;

  constructor() {
    this.baseUrl = process.env.REACT_APP_API_URL || 'http://localhost:8000';
    this.token = localStorage.getItem('token') || '';
  }

  private async fetchWithAuth<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<T> {
    const response = await fetch(`${this.baseUrl}${endpoint}`, {
      ...options,
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${this.token}`,
        ...options.headers,
      },
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.message || 'An error occurred');
    }

    return response.json();
  }

  async getCapabilities(): Promise<AgentCapabilities> {
    return this.fetchWithAuth<AgentCapabilities>('/api/agent/capabilities');
  }

  async generateCode(request: CodeGenerationRequest): Promise<CodeGenerationResponse> {
    return this.fetchWithAuth<CodeGenerationResponse>('/api/agent/generate', {
      method: 'POST',
      body: JSON.stringify(request),
    });
  }

  async getGenerationHistory(projectId: string): Promise<CodeGenerationResponse[]> {
    return this.fetchWithAuth<CodeGenerationResponse[]>(
      `/api/agent/generate/history/${projectId}`
    );
  }

  async reviewCode(request: CodeReviewRequest): Promise<CodeReviewResponse> {
    return this.fetchWithAuth<CodeReviewResponse>('/api/agent/review', {
      method: 'POST',
      body: JSON.stringify(request),
    });
  }

  async getReviews(projectId: string): Promise<CodeReviewResponse[]> {
    return this.fetchWithAuth<CodeReviewResponse[]>(
      `/api/agent/review/history/${projectId}`
    );
  }

  async generateTests(request: TestGenerationRequest): Promise<TestGenerationResponse> {
    return this.fetchWithAuth<TestGenerationResponse>('/api/agent/tests', {
      method: 'POST',
      body: JSON.stringify(request),
    });
  }

  async getTestSuites(projectId: string): Promise<TestGenerationResponse[]> {
    return this.fetchWithAuth<TestGenerationResponse[]>(
      `/api/agent/tests/history/${projectId}`
    );
  }

  async generateDocumentation(
    request: DocumentationRequest
  ): Promise<DocumentationResponse> {
    return this.fetchWithAuth<DocumentationResponse>('/api/agent/docs', {
      method: 'POST',
      body: JSON.stringify(request),
    });
  }

  async getDocumentation(projectId: string): Promise<DocumentationResponse[]> {
    return this.fetchWithAuth<DocumentationResponse[]>(
      `/api/agent/docs/history/${projectId}`
    );
  }

  async downloadGeneratedCode(id: string): Promise<Blob> {
    const response = await fetch(`${this.baseUrl}/api/agent/generate/${id}/download`, {
      headers: {
        Authorization: `Bearer ${this.token}`,
      },
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.message || 'Failed to download generated code');
    }

    return response.blob();
  }

  async downloadTestSuite(id: string): Promise<Blob> {
    const response = await fetch(`${this.baseUrl}/api/agent/tests/${id}/download`, {
      headers: {
        Authorization: `Bearer ${this.token}`,
      },
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.message || 'Failed to download test suite');
    }

    return response.blob();
  }

  async downloadDocumentation(id: string): Promise<Blob> {
    const response = await fetch(`${this.baseUrl}/api/agent/docs/${id}/download`, {
      headers: {
        Authorization: `Bearer ${this.token}`,
      },
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.message || 'Failed to download documentation');
    }

    return response.blob();
  }

  async cancelOperation(id: string): Promise<void> {
    await this.fetchWithAuth(`/api/agent/operations/${id}/cancel`, {
      method: 'POST',
    });
  }

  async getOperationStatus(id: string): Promise<{
    status: 'pending' | 'processing' | 'completed' | 'failed' | 'cancelled';
    progress?: number;
    error?: string;
  }> {
    return this.fetchWithAuth(`/api/agent/operations/${id}/status`);
  }
} 

File: ADE-stable-1.0/backend/services/agent_service.py
from typing import List, Optional, Dict, Any
from datetime import datetime
import uuid
from ..models.agent_capabilities import (
    GenerationType,
    GenerationContext,
    GeneratedCode,
    ReviewType,
    CodeReviewRequest,
    CodeReview,
    TestType,
    TestRequest,
    TestSuite,
    DocumentationType,
    DocumentationRequest,
    Documentation
)
from ..core.ai import AIService
from ..core.code_analysis import CodeAnalysisService
from ..core.security import get_current_user
from ..core.testing import TestService
from ..core.documentation import DocumentationService

class AgentService:
    def __init__(self):
        self.ai_service = AIService()
        self.code_analysis_service = CodeAnalysisService()
        self.test_service = TestService()
        self.documentation_service = DocumentationService()

    async def generate_code(self, context: GenerationContext, current_user: Any) -> GeneratedCode:
        """
        Generate code based on the provided context and requirements.
        This implementation uses AI to generate code and includes additional analysis.
        """
        # Validate project access
        if not await self._validate_project_access(context.project_id, current_user):
            raise ValueError("User does not have access to this project")

        # Generate code using AI
        generated_content = await self.ai_service.generate_code(
            language=context.language,
            framework=context.framework,
            requirements=context.requirements,
            constraints=context.constraints,
            style_guide=context.style_guide,
            existing_code=context.existing_code
        )

        # Analyze the generated code
        analysis_result = await self.code_analysis_service.analyze_code(
            project_id=context.project_id,
            content=generated_content,
            language=context.language
        )

        # Generate tests if needed
        tests = None
        if context.metadata and context.metadata.get("generate_tests", False):
            tests = await self._generate_tests_for_code(
                code=generated_content,
                language=context.language,
                framework=context.framework
            )

        # Generate documentation if needed
        documentation = None
        if context.metadata and context.metadata.get("generate_docs", False):
            documentation = await self._generate_documentation_for_code(
                code=generated_content,
                language=context.language,
                framework=context.framework
            )

        # Create the response
        return GeneratedCode(
            id=str(uuid.uuid4()),
            type=context.metadata.get("type", GenerationType.CODE),
            content=generated_content,
            language=context.language,
            file_path=context.metadata.get("file_path", "generated_code.py"),
            dependencies=analysis_result.get("dependencies"),
            tests=tests,
            documentation=documentation,
            generated_at=datetime.utcnow(),
            metadata={
                "analysis_result": analysis_result,
                "user_id": current_user.id,
                "project_id": context.project_id,
                **context.metadata
            }
        )

    async def _validate_project_access(self, project_id: str, current_user: Any) -> bool:
        """Validate that the current user has access to the project"""
        # TODO: Implement project access validation
        return True

    async def _generate_tests_for_code(
        self,
        code: str,
        language: str,
        framework: Optional[str] = None
    ) -> List[str]:
        """Generate tests for the given code"""
        # TODO: Implement test generation logic
        return []

    async def _generate_documentation_for_code(
        self,
        code: str,
        language: str,
        framework: Optional[str] = None
    ) -> str:
        """Generate documentation for the given code"""
        # TODO: Implement documentation generation logic
        return ""

    async def review_code(self, request: CodeReviewRequest, current_user: Any) -> CodeReview:
        """Perform code review"""
        # Validate project access
        if not await self._validate_project_access(request.project_id, current_user):
            raise ValueError("User does not have access to this project")

        # Get file content
        file_content = await self._get_file_content(request.project_id, request.file_path)
        
        # Perform code analysis
        analysis_result = await self.code_analysis_service.analyze_code(
            project_id=request.project_id,
            content=file_content,
            language=self._get_file_language(request.file_path)
        )

        # Generate review comments using AI
        comments = await self.ai_service.generate_review_comments(
            code=file_content,
            review_types=request.review_types,
            context=request.context,
            focus_areas=request.focus_areas,
            analysis_result=analysis_result
        )

        # Generate summary and recommendations
        summary, recommendations = await self.ai_service.generate_review_summary(
            comments=comments,
            analysis_result=analysis_result
        )

        return CodeReview(
            id=str(uuid.uuid4()),
            project_id=request.project_id,
            file_path=request.file_path,
            review_types=request.review_types,
            comments=comments,
            summary=summary,
            recommendations=recommendations,
            generated_at=datetime.utcnow(),
            metadata={
                "analysis_result": analysis_result,
                "user_id": current_user.id,
                "project_id": request.project_id,
                **request.metadata
            }
        )

    async def generate_tests(self, request: TestRequest, current_user: Any) -> TestSuite:
        """Generate test suite"""
        # Validate project access
        if not await self._validate_project_access(request.project_id, current_user):
            raise ValueError("User does not have access to this project")

        # Get file content
        file_content = await self._get_file_content(request.project_id, request.file_path)
        
        # Analyze code for test generation
        analysis_result = await self.code_analysis_service.analyze_code(
            project_id=request.project_id,
            content=file_content,
            language=self._get_file_language(request.file_path)
        )

        # Generate test cases using AI
        test_cases = await self.ai_service.generate_test_cases(
            code=file_content,
            test_types=request.test_types,
            coverage_target=request.coverage_target,
            framework=request.framework,
            analysis_result=analysis_result
        )

        # Generate test setup and teardown
        for test_case in test_cases:
            test_case.setup = await self.ai_service.generate_test_setup(
                test_case=test_case,
                framework=request.framework
            )
            test_case.teardown = await self.ai_service.generate_test_teardown(
                test_case=test_case,
                framework=request.framework
            )

        # Calculate coverage
        coverage = await self.test_service.calculate_coverage(
            code=file_content,
            test_cases=test_cases,
            framework=request.framework
        )

        return TestSuite(
            id=str(uuid.uuid4()),
            project_id=request.project_id,
            file_path=request.file_path,
            test_types=request.test_types,
            test_cases=test_cases,
            coverage=coverage,
            generated_at=datetime.utcnow(),
            metadata={
                "analysis_result": analysis_result,
                "user_id": current_user.id,
                "project_id": request.project_id,
                **request.metadata
            }
        )

    async def generate_documentation(self, request: DocumentationRequest, current_user: Any) -> Documentation:
        """Generate documentation"""
        # Validate project access
        if not await self._validate_project_access(request.project_id, current_user):
            raise ValueError("User does not have access to this project")

        # Get project structure and content
        project_structure = await self._get_project_structure(request.project_id)
        
        # Generate documentation content using AI
        content = await self.ai_service.generate_documentation(
            project_structure=project_structure,
            doc_types=request.doc_types,
            format=request.format,
            style=request.style,
            include_examples=request.include_examples
        )

        # Generate examples if requested
        examples = None
        if request.include_examples:
            examples = await self.ai_service.generate_documentation_examples(
                project_structure=project_structure,
                doc_types=request.doc_types
            )

        # Generate diagrams if needed
        if any(doc_type in request.doc_types for doc_type in [DocumentationType.ARCHITECTURE, DocumentationType.API]):
            diagrams = await self.documentation_service.generate_diagrams(
                project_structure=project_structure,
                doc_types=request.doc_types
            )
            content = await self.documentation_service.integrate_diagrams(
                content=content,
                diagrams=diagrams,
                format=request.format
            )

        return Documentation(
            id=str(uuid.uuid4()),
            project_id=request.project_id,
            doc_types=request.doc_types,
            content=content,
            format=request.format,
            examples=examples,
            generated_at=datetime.utcnow(),
            metadata={
                "user_id": current_user.id,
                "project_id": request.project_id,
                **request.metadata
            }
        )

    # Helper methods
    async def _get_file_content(self, project_id: str, file_path: str) -> str:
        """Get file content from project"""
        # TODO: Implement file content retrieval
        return ""

    async def _get_file_language(self, file_path: str) -> str:
        """Determine file language from extension"""
        # TODO: Implement language detection
        return "python"

    async def _get_project_structure(self, project_id: str) -> Dict[str, Any]:
        """Get project structure and content"""
        # TODO: Implement project structure retrieval
        return {} 